File: bugsnag-android-core/src/androidTest/java/com/bugsnag/android/SessionPayloadTest.java
package com.bugsnag.android;

import static com.bugsnag.android.BugsnagTestUtils.generateClient;
import static com.bugsnag.android.BugsnagTestUtils.generateSession;
import static com.bugsnag.android.BugsnagTestUtils.streamableToJson;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import android.content.Context;

import androidx.test.core.app.ApplicationProvider;

import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.util.Date;
import java.util.List;

public class SessionPayloadTest {

    private JSONObject rootNode;
    private Session session;
    private AppDataCollector appDataCollector;

    private SessionStore sessionStore;
    private File storageDir;
    private SessionPayload payload;
    private DeviceDataCollector deviceDataCollector;
    private Client client;

    /**
     * Configures a session tracking payload and session store, ensuring that 0 files are present
     *
     * @throws Exception if initialisation failed
     */
    @Before
    public void setUp() throws Exception {
        Context context = ApplicationProvider.getApplicationContext();
        sessionStore = new SessionStore(context, NoopLogger.INSTANCE, null);

        Assert.assertNotNull(sessionStore.storeDirectory);
        storageDir = new File(sessionStore.storeDirectory);
        FileUtils.clearFilesInDir(storageDir);
        session = generateSession();
        client = generateClient();
        payload = generatePayloadFromSession(context, session);
        rootNode = streamableToJson(payload);
    }

    private SessionPayload generatePayloadFromSession(Context context,
                                                      Session session) {
        appDataCollector = client.getAppDataCollector();
        deviceDataCollector = client.deviceDataCollector;
        return new SessionPayload(session, null, appDataCollector.generateApp(),
                deviceDataCollector.generateDevice());
    }

    /**
     * Deletes any files in the session store created during the test
     *
     * @throws Exception if the operation fails
     */
    @After
    public void tearDown() {
        FileUtils.clearFilesInDir(storageDir);
        client.close();
    }

    /**
     * Serialises sessions from a file instead
     */
    @Test
    public void testMultipleSessionFiles() throws Exception {
        sessionStore.write(session);
        sessionStore.write(generateSession());
        List<File> storedFiles = sessionStore.findStoredFiles();

        SessionPayload payload = new SessionPayload(null,
            storedFiles, appDataCollector.generateApp(), deviceDataCollector.generateDevice());
        rootNode = streamableToJson(payload);

        assertNotNull(rootNode);
        JSONArray sessions = rootNode.getJSONArray("sessions");
        assertNotNull(sessions);
        assertEquals(2, sessions.length());
    }

    @Test
    public void testAutoCapturedOverride() throws Exception {
        session = new Session("id", new Date(), null, false);
        Context context = ApplicationProvider.getApplicationContext();
        payload = generatePayloadFromSession(context, session);
        assertFalse(session.isAutoCaptured());
        session.setAutoCaptured(true);
        assertTrue(session.isAutoCaptured());

        JSONObject rootNode = streamableToJson(payload);
        JSONObject sessionNode = rootNode.getJSONArray("sessions").getJSONObject(0);
        assertFalse(sessionNode.has("user"));
    }
}

File: bugsnag-android-core/src/androidTest/java/com/bugsnag/android/SessionPayloadTest.kt
package com.bugsnag.android

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import java.io.BufferedWriter
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStreamWriter
import java.io.Writer
import java.util.Date

class SessionPayloadTest {
    lateinit var session: Session

    @Before
    fun setUp() {
        session = BugsnagTestUtils.generateSession()
        session.app = BugsnagTestUtils.generateApp()
        session.device = BugsnagTestUtils.generateDevice()
    }

    /**
     * Serialises sessions from a file instead
     */

    @Test
    fun testSessionV2PayloadFromFile() {
        // write file to disk

        val v2File = File.createTempFile(
            "150450000000053a27e4e-967c-4e5c-91be-2e86f2eb7cdc",
            "_v2.json"
        )
        val fos = FileOutputStream(v2File.absolutePath)
        val out: Writer = BufferedWriter(OutputStreamWriter(fos, "UTF-8"))
        val stream = JsonStream(out)
        session.toStream(stream)
        out.flush()
        val v2Payload = Session(v2File, Notifier(), NoopLogger, "_my-uuid-uuuuuuuuuuuuuuuuuuuuuuuuuuuu")
        val v2Obj = BugsnagTestUtils.streamableToJson(v2Payload)
        assertNotNull(v2Obj.getJSONObject("app"))
        assertNotNull(v2Obj.getJSONObject("device"))
        assertNotNull(v2Obj.getJSONObject("notifier"))
        val sessions = v2Obj.getJSONArray("sessions")
        assertNotNull(sessions)
        assertEquals(1, sessions.length().toLong())
        val session = sessions.getJSONObject(0)
        assertEquals("test", session.getString("id"))
        assertFalse(session.has("notifier"))
    }

    @Test
    fun testSessionV3PayloadFromFile() {
        // write file to disk

        val v3File = File.createTempFile(
            "150450000000053a27e4e-967c-4e5c-91be-2e86f2eb7cdc",
            "_v3.json"
        )
        val fos = FileOutputStream(v3File.absolutePath)
        val out: Writer = BufferedWriter(OutputStreamWriter(fos, "UTF-8"))
        val stream = JsonStream(out)
        session.toStream(stream)
        out.flush()
        val v3Payload = Session(v3File, Notifier(), NoopLogger, "_my-uuid-uuuuuuuuuuuuuuuuuuuuuuuuuuuu")
        val v3Object = BugsnagTestUtils.streamableToJson(v3Payload)
        assertNotNull(v3Object)
        assertNotNull(v3Object.getJSONObject("app"))
        assertNotNull(v3Object.getJSONObject("device"))
        assertNotNull(v3Object.getJSONObject("notifier"))
        val sessions = v3Object.getJSONArray("sessions")
        assertNotNull(sessions)
        assertEquals(1, sessions.length().toLong())
        val session = sessions.getJSONObject(0)
        assertEquals("test", session.getString("id"))
        assertFalse(session.has("notifier"))
    }

    @Test
    fun testAutoCapturedOverride() {
        session = Session(
            "id",
            Date(),
            null,
            false,
            Notifier(),
            NoopLogger,
            "TEST APIKEY"
        )
        assertFalse(session.isAutoCaptured)
        session.isAutoCaptured = true
        assertTrue(session.isAutoCaptured)
        val obj = BugsnagTestUtils.streamableToJson(session)
        val sessionNode = obj.getJSONArray("sessions").getJSONObject(0)
        assertFalse(sessionNode.has("user"))
    }
}
